File: java-tcp/replicant/src/main/java/multipaxos/MultiPaxos.java
package multipaxos;

import static multipaxos.network.ResponseType.OK;
import static multipaxos.network.ResponseType.REJECT;

import multipaxos.network.ChannelMap;
import multipaxos.network.Message.MessageType;
import multipaxos.network.PrepareRequest;
import multipaxos.network.PrepareResponse;
import multipaxos.network.AcceptRequest;
import multipaxos.network.AcceptResponse;
import multipaxos.network.CommitRequest;
import multipaxos.network.CommitResponse;

import ch.qos.logback.classic.Logger;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;
import log.Log;
import multipaxos.network.TcpLink;
import org.codehaus.jackson.map.ObjectMapper;
import org.slf4j.LoggerFactory;
import log.Instance;

class Peer {

  public final long id;
  public final TcpLink stub;

  public Peer(long id, TcpLink stub) {
    this.id = id;
    this.stub = stub;
  }
}

public class MultiPaxos {

  protected static final long kIdBits = 0xff;
  protected static final long kRoundIncrement = kIdBits + 1;
  protected static final long kMaxNumPeers = 0xf;
  private static final Logger logger = (Logger) LoggerFactory.getLogger(MultiPaxos.class);
  private AtomicLong ballot;
  private final Log log;
  private final long id;
  private final AtomicBoolean commitReceived;
  private final long commitInterval;
  private final int numPeers;
  private final List<Peer> peers;
  private final AtomicLong nextChannelId;
  private final ChannelMap channels;
  private final ReentrantLock mu;
  private final ExecutorService threadPool;

  private final Condition cvLeader;
  private final Condition cvFollower;

  private final AtomicBoolean prepareThreadRunning;
  private final ExecutorService prepareThread;

  private final AtomicBoolean commitThreadRunning;
  private final ExecutorService commitThread;

  public MultiPaxos(Log log, Configuration config) {
    this.ballot = new AtomicLong(kMaxNumPeers);
    this.log = log;
    this.id = config.getId();
    commitReceived = new AtomicBoolean(false);
    this.commitInterval = config.getCommitInterval();
    this.numPeers = config.getPeers().size();
    this.nextChannelId = new AtomicLong(0);
    this.channels = new ChannelMap();

    threadPool = Executors.newFixedThreadPool(config.getThreadPoolSize());
    prepareThreadRunning = new AtomicBoolean(false);
    commitThreadRunning = new AtomicBoolean(false);

    mu = new ReentrantLock();
    cvLeader = mu.newCondition();
    cvFollower = mu.newCondition();
    prepareThread = Executors.newSingleThreadExecutor();
    commitThread = Executors.newSingleThreadExecutor();

    peers = new ArrayList<>();
    long peerId = 0;
    for (var address : config.getPeers()) {
      peers.add(new Peer(peerId++, new TcpLink(address, channels)));
    }
  }

  public long ballot() {
    return ballot.get();
  }

  public long nextBallot() {
    long nextBallot = ballot.get();
    nextBallot += kRoundIncrement;
    nextBallot = (nextBallot & ~kIdBits) | id;
    return nextBallot;
  }

  public void becomeLeader(long newBallot, long newLastIndex) {
    mu.lock();
    try {
      logger.debug(id + " became a leader: ballot: " + ballot + " -> " + newBallot);
      ballot.set(newBallot);
      log.setLastIndex(newLastIndex);
      cvLeader.signal();
    } finally {
      mu.unlock();
    }
  }

  public void becomeFollower(long newBallot) {
    mu.lock();
    try {
      if (newBallot <= ballot.get()) {
        return;
      }
      var oldLeaderId = extractLeaderId(ballot.get());
      var newLeaderId = extractLeaderId(newBallot);
      if (newLeaderId != id && (oldLeaderId == id || oldLeaderId == kMaxNumPeers)) {
        logger.debug(id + " became a follower: ballot: " + ballot + " -> " + newBallot);
        cvFollower.signal();
      }
      ballot.set(newBallot);
    } finally {
      mu.unlock();
    }
  }

  public void sleepForCommitInterval() {
    try {
      Thread.sleep(commitInterval);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

  public void sleepForRandomInterval() {
    Random random = new Random();
    var sleepTime = random.nextInt(0, (int) commitInterval / 2);
    try {
      Thread.sleep(commitInterval + commitInterval / 2 + sleepTime);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

  public boolean receivedCommit() {
    return commitReceived.compareAndExchange(true, false);
  }

  void prepareThread() {
    while (prepareThreadRunning.get()) {
      mu.lock();
      try {
        while (prepareThreadRunning.get() && isLeader(this.ballot.get(), this.id)) {
          cvFollower.await();
        }
      } catch (InterruptedException e) {
        e.printStackTrace();
      } finally {
        mu.unlock();
      }
      while (prepareThreadRunning.get()) {
        sleepForRandomInterval();
        if (receivedCommit()) {
          continue;
        }
        var nextBallot = nextBallot();
        try {
          var r = runPreparePhase(nextBallot);
          if (r != null) {
            var maxLastIndex = r.getKey();
            var log = r.getValue();
            becomeLeader(nextBallot, maxLastIndex);
            replay(nextBallot, log);
            break;
          }
        } catch (IOException | InterruptedException e) {
          logger.error(e.getMessage());
        }
      }
    }
  }

  public void commitThread() {
    while (commitThreadRunning.get()) {
      mu.lock();
      try {
        while (commitThreadRunning.get() && !isLeader(this.ballot.get(), this.id)) {
          cvLeader.await();
        }
      } catch (InterruptedException e) {
        e.printStackTrace();
      } finally {
        mu.unlock();
      }
      var gle = log.getGlobalLastExecuted();
      while (commitThreadRunning.get()) {
        var ballot = ballot();
        if (!isLeader(ballot, this.id)) {
          break;
        }
        try {
          gle = runCommitPhase(ballot, gle);
        } catch (IOException | InterruptedException e) {
          logger.error(e.getMessage());
        }
        sleepForCommitInterval();
      }
    }
  }

  public Map.Entry<Long, HashMap<Long, log.Instance>> runPreparePhase(long ballot)
      throws IOException, InterruptedException {
    var numOks = 0;
    var log = new HashMap<Long, log.Instance>();
    long maxLastIndex = 0;

    if (ballot > this.ballot.get()) {
      numOks++;
      log = this.log.getLog();
      maxLastIndex = this.log.getLastIndex();
    } else {
      return null;
    }

    ObjectMapper mapper = new ObjectMapper();
    var request = mapper.writeValueAsString(new PrepareRequest(ballot, id));
    var channelId = nextChannelId.addAndGet(1);
    var responseChan = addChannel(channelId);

    for (var peer : peers) {
      if (peer.id != this.id) {
        threadPool.submit(() -> {
          peer.stub.sendAwaitResponse(MessageType.PREPAREREQUEST, channelId,
              request);
          logger.debug(id + " sent prepare request to " + peer.id);
        });
      }
    }

    while (true) {
      var response = responseChan.take();
      var prepareResponse = mapper.readValue(response, PrepareResponse.class);
      if (prepareResponse.getType() == OK) {
        numOks++;
        for (var instance : prepareResponse.getLogs()) {
          if (instance.getIndex() > maxLastIndex) {
            maxLastIndex = instance.getIndex();
          }
          Log.insert(log, instance);
        }
      } else {
        becomeFollower(prepareResponse.getBallot());
        break;
      }
      if (numOks > numPeers / 2) {
        removeChannel(channelId);
        return new HashMap.SimpleEntry<>(maxLastIndex, log);
      }
    }
    removeChannel(channelId);
    return null;
  }

  public Result runAcceptPhase(long ballot, long index, command.Command command,
      long clientId) {
    var numOks = 0;

    var instance = new log.Instance();
    if (ballot == this.ballot.get()) {
      numOks++;
      instance.setBallot(ballot);
      instance.setIndex(index);
      instance.setClientId(clientId);
      instance.setState(Instance.InstanceState.kInProgress);
      instance.setCommand(command);
      log.append(instance);
      if (numOks > numPeers / 2) {
        log.commit(index);
        return new Result(MultiPaxosResultType.kOk, (long) -1);
      }
    } else {
      var currentLeaderId = extractLeaderId(this.ballot.get());
      return new Result(MultiPaxosResultType.kSomeoneElseLeader,
          currentLeaderId);
    }

    ObjectMapper mapper = new ObjectMapper();
    String request;
    try {
      request = mapper.writeValueAsString(new AcceptRequest(id, instance));
    } catch (IOException e) {
      logger.error(e.getMessage());
      return new Result(MultiPaxosResultType.kRetry, null);
    }
    var channelId = nextChannelId.addAndGet(1);
    var responseChan = addChannel(channelId);

    for (var peer : peers) {
      if (peer.id != this.id) {
        threadPool.submit(() -> {
          peer.stub.sendAwaitResponse(MessageType.ACCEPTREQUEST, channelId,
              request);
          logger.debug(id + " sent accept request to " + peer.id);
        });
      }
    }
    while (true) {
      try {
        var response = responseChan.take();
        var acceptResponse = mapper.readValue(response, AcceptResponse.class);
        if (acceptResponse.getType() == OK) {
          numOks++;
        } else {
          becomeFollower(acceptResponse.getBallot());
          break;
        }
      } catch (InterruptedException | IOException e) {
        logger.error(e.getMessage());
      }
      if (numOks > numPeers / 2) {
        log.commit(index);
        removeChannel(channelId);
        return new Result(MultiPaxosResultType.kOk, null);
      }
    }
    removeChannel(channelId);
    if (!isLeader(this.ballot.get(), id)) {
      return new Result(MultiPaxosResultType.kSomeoneElseLeader,
          extractLeaderId(ballot()));
    }
    return new Result(MultiPaxosResultType.kRetry, null);
  }

  public Long runCommitPhase(long ballot, long globalLastExecuted)
      throws IOException, InterruptedException {
    var numOks = 0;
    var minLastExecuted = log.getLastExecuted();

    numOks++;
    log.trimUntil(globalLastExecuted);
    if (numOks == numPeers) {
      return minLastExecuted;
    }

    ObjectMapper mapper = new ObjectMapper();
    var request = mapper.writeValueAsString(
        new CommitRequest(ballot, minLastExecuted, globalLastExecuted, id));
    var channelId = nextChannelId.addAndGet(1);
    var responseChan = addChannel(channelId);

    for (var peer : peers) {
      if (peer.id != this.id) {
        threadPool.submit(() -> {
          peer.stub.sendAwaitResponse(MessageType.COMMITREQUEST, channelId,
              request);
          logger.debug(id + " sent commit to " + peer.id);
        });
      }
    }

    while (true) {
      var response = responseChan.take();
      var commitResponse = mapper.readValue(response, CommitResponse.class);
      if (commitResponse.getType() == OK) {
        numOks++;
        if (commitResponse.getLastExecuted() < minLastExecuted) {
          minLastExecuted = commitResponse.getLastExecuted();
        }
      } else {
        becomeFollower(commitResponse.getBallot());
        break;
      }
      if (numOks == numPeers) {
        removeChannel(channelId);
        return minLastExecuted;
      }
    }
    removeChannel(channelId);
    return globalLastExecuted;
  }

  public void replay(long ballot, HashMap<Long, log.Instance> log) {
    for (Map.Entry<Long, log.Instance> entry : log.entrySet()) {
      var r = runAcceptPhase(ballot, entry.getValue().getIndex(),
          entry.getValue().getCommand(),
          entry.getValue().getClientId());
      while (r.type == MultiPaxosResultType.kRetry) {
        r = runAcceptPhase(ballot, entry.getValue().getIndex(),
            entry.getValue().getCommand(),
            entry.getValue().getClientId());
      }
      if (r.type == MultiPaxosResultType.kSomeoneElseLeader) {
        return;
      }
    }
  }

  public BlockingQueue<String> addChannel(long channelId) {
    BlockingQueue<String> channel = new LinkedBlockingDeque<>(numPeers - 1);
    channels.put(channelId, channel);
    return channel;
  }

  public void removeChannel(long channelId) {
    channels.delete(channelId);
  }

  public void start() {
    startPrepareThread();
    startCommitThread();
  }

  public void stop() {
    stopPrepareThread();
    stopCommitThread();

    threadPool.shutdown();
    try {
      threadPool.awaitTermination(1000, TimeUnit.MILLISECONDS);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

  public void startPrepareThread() {
    logger.debug(id + " starting prepare thread");
    assert (!prepareThreadRunning.get());
    prepareThreadRunning.set(true);
    prepareThread.submit(this::prepareThread);
  }

  public void stopPrepareThread() {
    logger.debug(id + " stopping prepare thread");
    assert (prepareThreadRunning.get());
    mu.lock();
    prepareThreadRunning.set(false);
    cvFollower.signal();
    mu.unlock();
    prepareThread.shutdown();
    try {
      prepareThread.awaitTermination(1000, TimeUnit.MILLISECONDS);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

  public void startCommitThread() {
    logger.debug(id + " starting commit thread");
    assert (!commitThreadRunning.get());
    commitThreadRunning.set(true);
    commitThread.submit(this::commitThread);
  }

  public void stopCommitThread() {
    logger.debug(id + " stopping commit thread");
    assert (commitThreadRunning.get());
    mu.lock();
    commitThreadRunning.set(false);
    cvLeader.signal();
    mu.unlock();
    commitThread.shutdown();
    try {
      commitThread.awaitTermination(1000, TimeUnit.MILLISECONDS);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

  public Result replicate(command.Command command, long clientId) {
    var ballot = ballot();
    if (isLeader(ballot, this.id)) {
      return runAcceptPhase(ballot, log.advanceLastIndex(), command, clientId);

    }
    if (isSomeoneElseLeader(ballot, this.id)) {
      return new Result(MultiPaxosResultType.kSomeoneElseLeader,
          extractLeaderId(ballot));
    }
    return new Result(MultiPaxosResultType.kRetry, null);
  }

  public PrepareResponse prepare(PrepareRequest request) {
    logger.debug(id + " <-- prepare-- " + request.getSender());

    if (request.getBallot() > ballot.get()) {
      becomeFollower(request.getBallot());
      return new PrepareResponse(OK, ballot(), log.instances());
    } else {
      return new PrepareResponse(REJECT, ballot(), null);
    }
  }

  public AcceptResponse accept(AcceptRequest request) {
    logger.debug(this.id + " <--accept---  " + request.getSender());

    if (request.getInstance().getBallot() >= this.ballot()) {
      log.append(request.getInstance());
      if (request.getInstance().getBallot() > this.ballot.get()) {
        becomeFollower(request.getInstance().getBallot());
      }
      return new AcceptResponse(OK, ballot());
    }
    if (request.getInstance().getBallot() < this.ballot.get()) {
      return new AcceptResponse(REJECT, ballot());
    }
    return new AcceptResponse(OK, ballot());
  }

  public CommitResponse commit(CommitRequest request) {
    logger.debug(id + " <--commit--- " + request.getSender());

    if (request.getBallot() >= ballot()) {
      commitReceived.set(true);
      log.commitUntil(request.getLastExecuted(), request.getBallot());
      log.trimUntil(request.getGlobalLastExecuted());
      if (request.getBallot() > ballot.get()) {
        becomeFollower(request.getBallot());
      }
      return new CommitResponse(OK, ballot(), log.getLastExecuted());
    } else {
      return new CommitResponse(REJECT, ballot(), 0);
    }
  }

  public static long extractLeaderId(long ballot) {
    return ballot & kIdBits;
  }

  public static boolean isLeader(long ballot, long id) {
    return extractLeaderId(ballot) == id;
  }

  public static boolean isSomeoneElseLeader(long ballot, long id) {
    return !isLeader(ballot, id) && extractLeaderId(ballot) < kMaxNumPeers;
  }

  public long getId() {
    return id;
  }

  public List<Peer> getPeers() {
    return peers;
  }

  public long nextChannelId() {
    return this.nextChannelId.addAndGet(1);
  }
}

File: c++-tcp/multipaxos.cc
#include "multipaxos.h"
#include "json.h"

#include <chrono>
#include <thread>

using namespace std::chrono;

using moodycamel::BlockingReaderWriterQueue;
using nlohmann::json;

MultiPaxos::MultiPaxos(Log* log, 
                       json const& config, 
                       asio::io_context* io_context)
    : ballot_(kMaxNumPeers),
      log_(log),
      id_(config["id"]),
      commit_received_(false),
      commit_interval_(config["commit_interval"]),
      engine_(id_),
      port_(config["peers"][id_]),
      num_peers_(config["peers"].size()),
      next_channel_id_(0),
      thread_pool_(config["threadpool_size"]),
      prepare_thread_running_(false),
      commit_thread_running_(false) {
  int64_t id = 0;
  for (std::string const peer : config["peers"]) {
      peers_.emplace_back(id++, 
          std::make_shared<TcpLink>(peer, io_context));
  }
}

void MultiPaxos::PrepareThread() {
  while (prepare_thread_running_) {
    {
      std::unique_lock lock(mu_);
      while (prepare_thread_running_ && IsLeader(ballot_, id_))
        cv_follower_.wait(lock);
    }
    while (prepare_thread_running_) {
      SleepForRandomInterval();
      if (ReceivedCommit())
        continue;
      auto next_ballot = NextBallot();
      auto r = RunPreparePhase(next_ballot);
      if (r) {
        auto [max_last_index, log] = *r;
        BecomeLeader(next_ballot, max_last_index);
        Replay(next_ballot, log);
        break;
      }
    }
  }
}

void MultiPaxos::CommitThread() {
  while (commit_thread_running_) {
    {
      std::unique_lock lock(mu_);
      while (commit_thread_running_ && !IsLeader(ballot_, id_))
        cv_leader_.wait(lock);
    }
    auto gle = log_->GlobalLastExecuted();
    while (commit_thread_running_) {
      auto ballot = Ballot();
      if (!IsLeader(ballot, id_))
        break;
      gle = RunCommitPhase(ballot, gle);
      SleepForCommitInterval();
    }
  }
}

std::optional<
    std::pair<int64_t, std::unordered_map<int64_t, Instance>>>
MultiPaxos::RunPreparePhase(int64_t ballot) {
  size_t num_oks = 0;
  std::unordered_map<int64_t, Instance> log;
  int64_t max_last_index = 0;

  if (ballot > ballot_) {
    ++num_oks;
    log = log_->GetLog();
    max_last_index = log_->LastIndex();
    if (num_oks > num_peers_ / 2) {
      return std::make_pair(max_last_index, std::move(log));
    }
  } else {
    return std::nullopt;
  }

  PrepareRequest prepare_request(ballot, id_);
  json request = prepare_request;
  int64_t channel_id = NextChannelId();
  BlockingConcurrentQueue<std::string> channel(num_peers_ - 1);

  for (auto& peer : peers_) {
    if (peer.id_ != id_) {
      asio::post(thread_pool_, [this, &peer, request, channel_id, &channel] {
        peer.stub_->SendAwaitResponse(PREPAREREQUEST, channel_id, channel,
                                      request.dump());
        DLOG(INFO) << id_ << " sent prepare request to " << peer.id_;
      });
    }
  }

  while (true) {
    std::string response;
    channel.wait_dequeue_timed(response, std::chrono::milliseconds(500));
    if (response.size() == 0) {
      break;
    }
    json prepare_response = json::parse(response);
    if (prepare_response["type_"] == OK) {
      ++num_oks;
      auto instances = prepare_response["instances_"]
          .get<std::vector<Instance>>();
      for (auto instance : instances) {
        max_last_index = std::max(max_last_index, instance.index_);
        Insert(&log, std::move(instance));
      }
    } else {
      BecomeFollower(prepare_response["ballot_"]);
      break;
    }
    if (num_oks > num_peers_ / 2) {
      // RemoveChannel(channel_id);
      return std::make_pair(max_last_index, std::move(log));
    }
  }
  // RemoveChannel(channel_id);
  return std::nullopt;
}

Result MultiPaxos::RunAcceptPhase(int64_t ballot,
                                  int64_t index,
                                  Command command,
                                  int64_t client_id) {
  size_t num_oks = 0;

  Instance instance(ballot, index, client_id, INPROGRESS, command);

  if (ballot == ballot_) {
    ++num_oks;
    log_->Append(instance);
    if (num_oks > num_peers_ / 2) {
      log_->Commit(index);
      return Result{ResultType::kOk, std::nullopt};
    }
  } else {
    auto leader = ExtractLeaderId(ballot_);
    return Result{ResultType::kSomeoneElseLeader, leader};
  }

  AcceptRequest accept_request(instance, id_);
  json request = accept_request;
  int64_t channel_id = NextChannelId();
  BlockingConcurrentQueue<std::string> channel(num_peers_ - 1);

  for (auto& peer : peers_) {
    if (peer.id_ != id_) {
      asio::post(thread_pool_, [this, &peer, request, channel_id, &channel] {
        peer.stub_->SendAwaitResponse(ACCEPTREQUEST, channel_id, channel,
                                      request.dump());
        DLOG(INFO) << id_ << " sent accept request to " << peer.id_;
      });
    }
  }

  while (true) {
    std::string response;
    channel.wait_dequeue_timed(response, std::chrono::milliseconds(3000));
    if (response.size() == 0) {
      break;
    }
    json accept_response = json::parse(response);
    if (accept_response["type_"] == OK)
      ++num_oks;
    else {
      BecomeFollower(accept_response["ballot_"]);
      break;
    }
    if (num_oks > num_peers_ / 2) {
      log_->Commit(index);
      // RemoveChannel(channel_id);
      return Result{ResultType::kOk, std::nullopt};
    }
  }
  // RemoveChannel(channel_id);
  if (!IsLeader(ballot_, id_))
    return Result{ResultType::kSomeoneElseLeader, ExtractLeaderId(ballot_)};
  return Result{ResultType::kRetry, std::nullopt};
}

int64_t MultiPaxos::RunCommitPhase(int64_t ballot,
                                   int64_t global_last_executed) {
  size_t num_oks = 0;
  int64_t min_last_executed = log_->LastExecuted();

  ++num_oks;
  log_->TrimUntil(global_last_executed);
  if (num_oks == num_peers_)
    return min_last_executed;

  CommitRequest commit_request(ballot, min_last_executed, global_last_executed, id_);
  json request = commit_request;
  int64_t channel_id = NextChannelId();
  BlockingConcurrentQueue<std::string> channel(num_peers_ - 1);

  for (auto& peer : peers_) {
    if (peer.id_ != id_) {
      asio::post(thread_pool_, [this, &peer, request, channel_id, &channel] {
        peer.stub_->SendAwaitResponse(COMMITREQUEST, channel_id, channel,
                                      request.dump());
        DLOG(INFO) << id_ << " sent commit to " << peer.id_;
      });
    }
  }

  while (true) {
    std::string response;
    channel.wait_dequeue_timed(response, std::chrono::milliseconds(commit_interval_));
    if (response.size() == 0) {
      break;
    }
    json commit_response = json::parse(response);
    if (commit_response["type_"] == OK) {
      ++num_oks;
      if (commit_response["last_executed_"] < min_last_executed)
        min_last_executed = commit_response["last_executed_"];
    } else {
      BecomeFollower(commit_response["ballot_"]);
      break;
    }
    if (num_oks == num_peers_) {
      // RemoveChannel(channel_id);
      return min_last_executed;
    }
  }
  // RemoveChannel(channel_id);
  return global_last_executed;
}

void MultiPaxos::Replay(
    int64_t ballot,
    std::unordered_map<int64_t, Instance> const& log) {
  for (auto const& [index, instance] : log) {
    Result r = RunAcceptPhase(ballot, instance.index_, instance.command_,
                              instance.client_id_);
    while (r.type_ == ResultType::kRetry)
      r = RunAcceptPhase(ballot, instance.index_, instance.command_,
                         instance.client_id_);
    if (r.type_ == ResultType::kSomeoneElseLeader)
      return;
  }
}

// BlockingReaderWriterQueue<std::string> 
//     MultiPaxos::AddChannel(int64_t channel_id) {
//   BlockingReaderWriterQueue<std::string> response_chan(num_peers_ - 1);
//   std::unique_lock lock(channels_.mu_);
//   auto [it, ok] = channels_.map_.insert({channel_id, response_chan});
//   CHECK(ok);
//   return response_chan;
// }

// void MultiPaxos::RemoveChannel(int64_t channel_id) {
//   std::unique_lock lock(channels_.mu_);
//   auto it = channels_.map_.find(channel_id);
//   CHECK(it != channels_.map_.end());
//   channels_.map_.erase(it);
// }

int64_t MultiPaxos::NextChannelId() {
  return ++next_channel_id_;
}

void MultiPaxos::Start() {
  StartPrepareThread();
  StartCommitThread();
}

void MultiPaxos::Stop() {
  StopPrepareThread();
  StopCommitThread();
  StopStub();
  thread_pool_.join();
}

void MultiPaxos::StartPrepareThread() {
  DLOG(INFO) << id_ << " starting prepare thread";
  CHECK(!prepare_thread_running_);
  prepare_thread_running_ = true;
  prepare_thread_ = std::thread(&MultiPaxos::PrepareThread, this);
}

void MultiPaxos::StopPrepareThread() {
  DLOG(INFO) << id_ << " stopping prepare thread";
  CHECK(prepare_thread_running_);
  {
    std::unique_lock lock(mu_);
    prepare_thread_running_ = false;
  }
  cv_follower_.notify_one();
  prepare_thread_.join();
}

void MultiPaxos::StartCommitThread() {
  DLOG(INFO) << id_ << " starting commit thread";
  CHECK(!commit_thread_running_);
  commit_thread_running_ = true;
  commit_thread_ = std::thread(&MultiPaxos::CommitThread, this);
}

void MultiPaxos::StopCommitThread() {
  DLOG(INFO) << id_ << " stopping commit thread";
  CHECK(commit_thread_running_);
  {
    std::unique_lock lock(mu_);
    commit_thread_running_ = false;
  }
  cv_leader_.notify_one();
  commit_thread_.join();
}

void MultiPaxos::StopStub() {
  for (auto& peer : peers_)
    peer.stub_->Stop();
}

Result MultiPaxos::Replicate(Command command, int64_t client_id) {
  auto ballot = Ballot();
  if (IsLeader(ballot, id_))
    return RunAcceptPhase(ballot, log_->AdvanceLastIndex(), std::move(command),
                          client_id);
  if (IsSomeoneElseLeader(ballot, id_))
    return Result{ResultType::kSomeoneElseLeader, ExtractLeaderId(ballot)};
  return Result{ResultType::kRetry, std::nullopt};
}

PrepareResponse MultiPaxos::Prepare(json const& request) {
  DLOG(INFO) << id_ << " <--prepare-- " << request["sender_"];
  int64_t ballot = request.at("ballot_").get<int64_t>();
  if (ballot > ballot_) {
    BecomeFollower(request["ballot_"]);
    PrepareResponse response(OK, ballot_, log_->Instances());
    return response;
  } else {
    PrepareResponse response(REJECT, ballot_, std::vector<Instance>());
    return response;
  }
}

AcceptResponse MultiPaxos::Accept(json const& request) {
  DLOG(INFO) << id_ << " <--accept--- " << request["sender_"];
  int64_t ballot = request.at("instance_").at("ballot_").get<int64_t>();
  if (ballot >= ballot_) {
    auto instance = request.at("instance_").get<Instance>();
    log_->Append(instance);
    if (ballot > ballot_)
      BecomeFollower(request["instance_"]["ballot_"]);
    AcceptResponse response(OK, ballot_);
    return response;
  }
  if (ballot < ballot_) {
    AcceptResponse response(REJECT, ballot_);
    return response;
  }
  AcceptResponse response(OK, ballot_);
  return response;
}

CommitResponse MultiPaxos::Commit(json const& request) {
  DLOG(INFO) << id_ << " <--commit--- " << request.at("sender_");
  int64_t ballot = request.at("ballot_").get<int64_t>();
  if (ballot >= ballot_) {
    commit_received_ = true;
    log_->CommitUntil(request["last_executed_"], request["ballot_"]);
    log_->TrimUntil(request["global_last_executed_"]);
    if (ballot > ballot_)
      BecomeFollower(request["ballot_"]);
    CommitResponse response(OK, ballot_, log_->LastExecuted());
    return response;
  } else {
    CommitResponse response(REJECT, ballot_, 0);
    return response;
  }
}
